File: SystemicRisk/core/template/fun_model_skeleton_template.py
"函数：通用模型框架"

## 函数：通用模型框架

##########################################
#状态/开发
##########################################

"""
通用模型框架模板：

Argument: 
- A:SystemicRiskAgent: Agent群变量；
- para:dict: 参数变量；
- env:dict: 环境变量；
- model:ModelComponent: 模型组件实例；
- A_data:AgentDataCollection: Agent群变量之数据；

Return:
- A:SystemicRiskAgent: Agent群变量；
- para:dict: 参数变量；
- env:dict: 环境变量；
- A_data:AgentDataCollection: Agent群变量之数据；
"""
def fun_model_skeleton_template(self, A:SystemicRiskAgent, para:dict, env:dict, model:ModelComponent, A_data:AgentDataCollection):
    if (env['is_model'])
        if (env['tau'] > 1)
            # @testprintln "\n继续模型：$(env['model_name'])"
        else:
            # @testprintln "\n开始模型：$(env['model_name'])"
            pass
        pass

    ## 运行每个过程
    for (idx_process, process) in enumerate(model.content)
        env['index_process'] = idx_process
        if env['index_process'] == env['loadedIndexProcess']: # 调度读取：如果当前过程等于待读取的过程，则进入继续读取。
            env['process_name'] = Symbol(process.functionName)
            A, para, env, A_data = runProcess(A, para, env, process, A_data)
            pass
        pass # for

    ## 判断是否结束步进
    if !env['is_step']
        # @testprintln "步进已结束，跳出模型：$(env['model_name'])。"
        pass
    # 判断是否结束模型
    if env['state_of_schedule'] == StateOfSchedule.idle
        env['is_model'] = False
        env['is_experiment'] = False
        pass
    if (!env['is_model'] || !env['is_experiment'])
        # @testprintln "结束模型：$(env['model_name'])。\n"
        pass


    # return BB, BI, para, env
    return A, para, env, A_data
    pass # functions

#     pass # module

File: draft/fun_model_skeleton_template.jl
"函数：通用模型框架"

## 函数：通用模型框架

##########################################
#状态/暂缓开发
##########################################

"""
通用模型框架：
Argument: 
- BB::BankCommercial: 商业银行群变量；
- BI::BankInterbank: 银行间邻接矩阵变量；
- para::Dict: 参数变量；
- env::Dict: 环境变量；
"""
function fun_model_skeleton!(BB::BankCommercial, BI::BankInterbank, para::Dict, env::Dict) # 此处需要修改函数名称为实际待生成函数模型名称

    env[:index_process] = 0 # 初始化过程所在位置
    # env[:state_of_schedule] = :stepping
    # @testprintln "切换调度运作状态为$(env[:state_of_schedule])"

    env[:tau] = 0 # 初始化回合

    if (env[:is_model])
        if (env[:tau] > 0)
            @testprintln "继续模型model：\n"
        else
            @testprintln "开始模型model：\n"
        end
    end

    ## 过程

    #=【插入表达式】=#

    ## 判断是否结束
    if !env[:is_step]
        @testprintln "步进已结束，跳出$(env[:model_name])。"
    end

    if env[:state_of_schedule] == :idle
        env[:is_model] = false
        env[:is_experiment] = false
    end

    if (!env[:is_model] || !env[:is_experiment])
        @testprintln "$(env[:model_name])结束。"
    end


    # return BB, BI, para, env
    # return BB, BI, BB_tau, BI_tau, para, env
end # function

# end # module






